tables/AlternateMans/BufferedColumnarFile.h

```diff
@@ -0,0 +1,307 @@
+   */
+  void Write(uint64_t row, uint64_t column_offset,
+             const std::complex<double>* data, uint64_t n) {
+    WriteImplementation(row, column_offset, data, n);
+  }
+  /**
+   * Write an array of bools. Bools are stored with bit-packing. See float
+   * version for documentation.
+   */
+  void Write(uint64_t row, uint64_t column_offset, const bool* data,
+             uint64_t n) {
+    assert(column_offset + n <= Stride());
+    ActivateBlock(row);
+    PackBoolArray(packed_buffer_.data(), data, n);
+    Seek(row * Stride() + column_offset + DataLocation(), SEEK_SET);
+    WriteData(packed_buffer_.data(), n);
+    SetNRows(std::max(row + 1, NRows()));
+  }
+
+  /**
+   * Set the number of bytes per row for this file. This changes the format
+   * of the file, and because of this the file is emptied.
+   */
+  void SetStride(uint64_t new_stride) {
+    RowBasedFile::SetStride(new_stride);
+    packed_buffer_.resize((new_stride + 7) / 8);
+    active_block_ = std::numeric_limits<uint64_t>::max();
+    rows_per_block_ =
+        new_stride == 0 ? 0 : std::max<size_t>(1, BufferSize / new_stride);
+    block_buffer_.resize(rows_per_block_ * new_stride);
+    block_changed_ = false;
+  }
+
+ private:
+  // Create or overwrite a new columnar file on disk
+  VarBufferedColumnarFile(const std::string& filename, uint64_t header_size,
+                          uint64_t stride)
+      : RowBasedFile(filename, header_size, stride),
+        packed_buffer_((stride + 7) / 8),
+        rows_per_block_(std::max<size_t>(1, BufferSize / stride)),
+        block_buffer_(rows_per_block_ * stride) {}
+
+  // Open an existing columnar file
+  VarBufferedColumnarFile(const std::string& filename, size_t header_size)
+      : RowBasedFile(filename, header_size) {
+    if (Stride() != 0) {
+      packed_buffer_.resize((Stride() + 7) / 8);
+      rows_per_block_ = std::max<size_t>(1, BufferSize / Stride());
+      block_buffer_.resize(rows_per_block_ * Stride());
+    }
+    active_block_ = std::numeric_limits<uint64_t>::max();
+  }
+
+  void ActivateBlock(uint64_t row) {
+    const uint64_t block = row / rows_per_block_;
+    if (active_block_ != block) {
+      if (block_changed_) {
+        const uint64_t start_row = active_block_ * rows_per_block_;
+        const size_t n_rows_to_write =
+            std::min(rows_per_block_, std::max(NRows(), start_row) - start_row);
+        Seek(start_row * Stride() + DataLocation(), SEEK_SET);
+        WriteData(block_buffer_.data(), n_rows_to_write * Stride());
+        block_changed_ = false;
+      }
+
+      const uint64_t start_row = block * rows_per_block_;
+      const size_t n_rows_to_read =
+          std::min(rows_per_block_, std::max(NRows(), start_row) - start_row);
+      if (n_rows_to_read > 0) {
+        Seek(start_row * Stride() + DataLocation(), SEEK_SET);
+        ReadData(block_buffer_.data(), n_rows_to_read * Stride());
+      }
```

Comment:
-  I don't see when `n_rows_to_read` can be zero here.
-  Even if it can become zero, the `if` statement is not needed:`ReadData` will just read nothing in that corner case.
```suggestion
      Seek(start_row * Stride() + DataLocation(), SEEK_SET);
      ReadData(block_buffer_.data(), n_rows_to_read * Stride());
```

Reply:
It will be zero when `NRows() <= start_row`, which happens when creating a file and e.g. activating the first block (NRows() == 0, start_row = 0), or a block that is not existing yet.

You're right the if is not strictly necessary, which I hadn't realized. It may be an optimization though; it prevents calling Seek() everytime a new block is started (so when writing a file from start to end, for all blocks). Even though a seek is a relatively expensive operation (it's a kernel call, but might even perform I/O), with a large buffer size, that overhead is relatively small though, but in the write case it is pure overhead: the `if` statement never evaluates to false. However, when reading an existing file, it returns false every time, so there the if is always overhead. I think I would indeed then just leave the if out as you suggest, as on average it won't help much, if any. So, I'll remove it.